diff --git a/BUILD b/BUILD
index 086ea605..943b0020 100644
--- a/BUILD
+++ b/BUILD
@@ -300,6 +300,7 @@ cc_library(
     srcs = ["mhlo/IR/hlo_ops_common.cc"],
     hdrs = ["mhlo/IR/hlo_ops_common.h"],
     includes = [
+        ".",
         "include",
     ],
     strip_include_prefix = ".",
@@ -403,6 +404,7 @@ cc_library(
         "lhlo/IR/lhlo_structured_interface.h.inc",
     ],
     includes = [
+        ".",
         "include",
     ],
     strip_include_prefix = ".",
@@ -418,6 +420,7 @@ cc_library(
     srcs = ["utils/convert_op_folder.cc"],
     hdrs = ["utils/convert_op_folder.h"],
     includes = [
+        ".",
         "include",
     ],
     strip_include_prefix = ".",
@@ -448,6 +451,7 @@ cc_library(
         "utils/hlo_utils.h",
     ],
     includes = [
+        ".",
         "include",
     ],
     strip_include_prefix = ".",
@@ -496,6 +500,7 @@ cc_library(
         "lhlo/utils/lhlo_utils.h",
     ],
     includes = [
+        ".",
         "include",
     ],
     strip_include_prefix = ".",
@@ -522,6 +527,7 @@ cc_library(
     srcs = ["lhlo_gpu/IR/lhlo_gpu_ops.cc"],
     hdrs = ["lhlo_gpu/IR/lhlo_gpu_ops.h"],
     includes = [
+        ".",
         "include",
     ],
     strip_include_prefix = ".",
@@ -545,6 +551,7 @@ cc_library(
     srcs = ["lhlo_gpu/IR/lhlo_gpu_ops.cc.inc"],
     hdrs = ["lhlo_gpu/IR/lhlo_gpu_ops.h.inc"],
     includes = [
+        ".",
         "include",
     ],
     strip_include_prefix = ".",
@@ -636,6 +643,7 @@ cc_library(
         "mhlo/utils/legalize_to_linalg_utils.h",
     ],
     includes = [
+        ".",
         "include",
     ],
     strip_include_prefix = ".",
@@ -695,7 +703,10 @@ cc_library(
     name = "type_conversion",
     srcs = ["mhlo/utils/type_conversion.cc"],
     hdrs = ["mhlo/utils/type_conversion.h"],
-    includes = ["include"],
+    includes = [
+        ".",
+        "include",
+    ],
     strip_include_prefix = ".",
     deps = [
         ":mlir_hlo",
@@ -829,6 +840,7 @@ cc_library(
     srcs = ["utils/codegen_utils.cc"],
     hdrs = ["utils/codegen_utils.h"],
     includes = [
+        ".",
         "include",
     ],
     strip_include_prefix = ".",
@@ -847,6 +859,7 @@ cc_library(
     name = "placement_utils",
     hdrs = ["utils/placement_utils.h"],
     includes = [
+        ".",
         "include",
     ],
     deps = ["@llvm-project//llvm:Support"],
@@ -1242,6 +1255,7 @@ cc_library(
     ],
     hdrs = ["gml_st/transforms/test_passes.h"],
     includes = [
+        ".",
         "include",
     ],
     strip_include_prefix = ".",
@@ -1300,7 +1314,7 @@ cc_library(
     name = "userange_analysis",
     srcs = ["lib/Analysis/userange_analysis.cc"],
     hdrs = ["include/mlir-hlo/Analysis/userange_analysis.h"],
-    includes = ["include"],
+    includes = [".", "include"],
     deps = [
         "@llvm-project//llvm:Support",
         "@llvm-project//mlir:Analysis",
@@ -1314,7 +1328,7 @@ cc_library(
     name = "shape_component_analysis",
     srcs = ["lib/Analysis/shape_component_analysis.cc"],
     hdrs = ["include/mlir-hlo/Analysis/shape_component_analysis.h"],
-    includes = ["include"],
+    includes = [".", "include"],
     deps = [
         ":mlir_hlo",
         "@llvm-project//llvm:Support",
@@ -1442,6 +1456,7 @@ cc_library(
     name = "CAPIHeaders",
     hdrs = CAPI_HEADERS,
     includes = [
+        ".",
         "include",
     ],
     strip_include_prefix = ".",
@@ -1589,6 +1604,7 @@ cc_library(
     srcs = ["gml_st/IR/gml_st_ops.cc"],
     hdrs = ["gml_st/IR/gml_st_ops.h"],
     includes = [
+        ".",
         "include",
     ],
     strip_include_prefix = ".",
@@ -1647,6 +1663,7 @@ cc_library(
     ],
     hdrs = ["gml_st/interfaces/tiling_interface.h"],
     includes = [
+        ".",
         "include",
     ],
     strip_include_prefix = ".",
@@ -1663,6 +1680,7 @@ cc_library(
     srcs = ["gml_st/interfaces/tiling_interface_impl.cc"],
     hdrs = ["gml_st/interfaces/tiling_interface_impl.h"],
     includes = [
+        ".",
         "include",
     ],
     strip_include_prefix = ".",
@@ -1690,6 +1708,7 @@ cc_library(
     srcs = ["gml_st/interfaces/bufferizable_op_interface_impl.cc"],
     hdrs = ["gml_st/interfaces/bufferizable_op_interface_impl.h"],
     includes = [
+        ".",
         "include",
     ],
     strip_include_prefix = ".",
@@ -1726,6 +1745,7 @@ cc_library(
     srcs = ["gml_st/transforms/transforms.cc"],
     hdrs = ["gml_st/transforms/transforms.h"],
     includes = [
+        ".",
         "include",
     ],
     strip_include_prefix = ".",
@@ -1790,6 +1810,7 @@ cc_library(
     srcs = ["thlo/IR/thlo_ops.cc"],
     hdrs = ["thlo/IR/thlo_ops.h"],
     includes = [
+        ".",
         "include",
     ],
     strip_include_prefix = ".",
@@ -1821,6 +1842,7 @@ cc_library(
     srcs = ["thlo/interfaces/bufferizable_op_interface_impl.cc"],
     hdrs = ["thlo/interfaces/bufferizable_op_interface_impl.h"],
     includes = [
+        ".",
         "include",
     ],
     strip_include_prefix = ".",
diff --git a/stablehlo/BUILD b/stablehlo/BUILD
index 3ab1b319..d11ba9d5 100644
--- a/stablehlo/BUILD
+++ b/stablehlo/BUILD
@@ -1,16 +1,16 @@
-load("//tools/build_defs/license:license.bzl", "license")
+# load("//tools/build_defs/license:license.bzl", "license")
 load("@llvm-project//mlir:tblgen.bzl", "gentbl_cc_library", "gentbl_filegroup", "td_library")
 
 package(
-    default_applicable_licenses = ["//stablehlo:license"],
+    # default_applicable_licenses = ["//stablehlo:license"],
     default_visibility = ["//visibility:public"],
     licenses = ["notice"],
 )
 
-license(
-    name = "license",
-    package_name = "stablehlo",
-)
+# license(
+#     name = "license",
+#     package_name = "stablehlo",
+# )
 
 exports_files([
     "LICENSE",
