diff --git a/lib/SPIRV/SPIRVInternal.h b/lib/SPIRV/SPIRVInternal.h
index 94a088f0..69403106 100644
--- a/lib/SPIRV/SPIRVInternal.h
+++ b/lib/SPIRV/SPIRVInternal.h
@@ -179,11 +179,12 @@ typedef SPIRVMap<Op, Op, IntBoolOpMapId> IntBoolOpMap;
   "-v512:512:512-v1024:1024:1024"
 
 enum SPIRAddressSpace {
-  SPIRAS_Private,
+  SPIRAS_Generic,
   SPIRAS_Global,
-  SPIRAS_Constant,
+  SPIRAS_Internal,
   SPIRAS_Local,
-  SPIRAS_Generic,
+  SPIRAS_Constant,
+  SPIRAS_Private,
   SPIRAS_GlobalDevice,
   SPIRAS_GlobalHost,
   SPIRAS_Input,
diff --git a/lib/SPIRV/SPIRVRegularizeLLVM.cpp b/lib/SPIRV/SPIRVRegularizeLLVM.cpp
index e9a42beb..9f550bde 100644
--- a/lib/SPIRV/SPIRVRegularizeLLVM.cpp
+++ b/lib/SPIRV/SPIRVRegularizeLLVM.cpp
@@ -426,7 +426,6 @@ bool SPIRVRegularizeLLVMBase::regularize() {
           FI->dropAllReferences();
           ToErase.push_back(FI);
         }
-
         // Remove metadata not supported by SPIRV
         static const char *MDs[] = {
             "fpmath",
diff --git a/lib/SPIRV/SPIRVWriter.cpp b/lib/SPIRV/SPIRVWriter.cpp
index 90b8c4e2..b049b125 100644
--- a/lib/SPIRV/SPIRVWriter.cpp
+++ b/lib/SPIRV/SPIRVWriter.cpp
@@ -367,12 +367,6 @@ SPIRVType *LLVMToSPIRVBase::transType(Type *T) {
   if (T->isArrayTy()) {
     // SPIR-V 1.3 s3.32.6: Length is the number of elements in the array.
     //                     It must be at least 1.
-    if (T->getArrayNumElements() < 1) {
-      std::string Str;
-      llvm::raw_string_ostream OS(Str);
-      OS << *T;
-      SPIRVCK(T->getArrayNumElements() >= 1, InvalidArraySize, OS.str());
-    }
     Type *ElTy = T->getArrayElementType();
     SPIRVType *TransType = BM->addArrayType(
         transType(ElTy),
